### ISSUE TITLE
cause ECONNRESET when proxy get request with body

### ISSUE THREAD

**Issue by vincent178** [fastify/fastify#953]
Hi,

I found an issue when using `fastify-reply-from` plugin. I proxy a `GET` request to a target server with message body (I input the body by mistake, but I think http allow us to send get request with body).

But fastify don't parse the body when it's a GET request, source code: https://github.com/fastify/fastify/blob/c21a45f29e0fe55b4f025519a21c2683f97c4969/lib/handleRequest.js#L17-L20

For `fastify-reply-from` plugin, we directly proxy the headers, which contains the content-length but without body (this body is from fastify).

It will trigger http_parser error, the behavior is that server will hang for around 2mins, and return error
```
socket hang up    
  at createHangUpError (_http_client.js:331:15)    
  at TLSSocket.socketOnEnd (_http_client.js:423:23)    
  at emitNone (events.js:111:20)    
  at TLSSocket.emit (events.js:208:7)    
  at endReadableNT (_stream_readable.js:1055:12)    
  at _combinedTickCallback (internal/process/next_tick.js:138:11)    
  at process._tickCallback (internal/process/next_tick.js:180:9
```

I could think of two ways to solve this, 
one, force fastify to parse GET request body, so when proxy, it will get the right body and with right content-length
two, force `fastify-reply-from` to reset the content-length when it's a `GET` request.

both have down sides:
first one, it will hurt performance a bit
second one, it seems a hack to me, 
I prefer the first one, it seems a correct way.

Could you please suggest a way to solve this?

Example code, with `fastify-reply-from`
```
const Fastify = require('fastify')

const proxy = Fastify({
  logger: true
})

proxy.register(require('fastify-reply-from'), {
  base: 'https://ruby-china.org'
});

proxy.all('/*', (request, reply) => reply.from(request.req.url));

proxy.listen(4000, (err) => {
  if (err) throw err;
});
```

use plain node to reproduce this issue:
```
const http = require('http');
const https = require('https');

const server = http.createServer((req, res) => {
  const proxyReq = https.request({
    method: 'GET',
    port: '',
    hostname: 'ruby-china.org', 
    pathname: '/api/v3/topics',
    headers: {
      host: "ruby-china.org", 
      "user-agent": "insomnia/5.16.4", 
      "content-type": "application/json",
      "content-length": "35", // the key is here, it's a wrong number
      "host": "ruby-china.org"
    },
  })
  proxyReq.end("");
});


server.listen(4001);
```

### Context
OS: MacOS 10.13
Node: 8.9.4

Thanks,
Vincent

**Comment by mcollina** [comment]
I think the example is missing the configuration of fastify-reply-from. Can you please add it?

**Comment by vincent178** [comment]
@mcollina sure, added, to use this example, send a http `GET` request with a message body.

**Comment by jsumners** [comment]
https://tools.ietf.org/html/rfc2616#section-9.3 Would mention at least in passing if a body should be expected during a GET request. Since it doesn't, we can assume that one should not be present.

**Comment by mcollina** [comment]
I think this is a bug in fastify-reply-from, and the content-length needs to be reset if it's a GET response. Or we might want for `fastify-reply-from` to throw an error.

I'm also ok in returning a 400 status code in fastify if there is a Content-Length property != 0 in a http GET.

**Comment by mcollina** [comment]
@vincent178 would you like to send a PR to fastify-reply-from?

**Comment by vincent178** [comment]
I think at least we should not return a 400 in `fastify`. According to https://tools.ietf.org/html/rfc2616#section-4.3

```
  The presence of a message-body in a request is signaled by the
   inclusion of a Content-Length or Transfer-Encoding header field in
   the request's message-headers. A message-body MUST NOT be included in
   a request if the specification of the request method (section 5.1.1)
   does not allow sending an entity-body in requests. A server SHOULD
   read and forward a message-body on any request;if the request method
   does not include defined semantics for an entity-body, then the
   message-body SHOULD be ignored when handling the request.
```
any thoughts?

@jsumners @mcollina

**Comment by jsumners** [comment]
Read the last clause in your quote:

> ... if the request method
   does not include defined semantics for an entity-body, then the
   message-body SHOULD be ignored when handling the request.

**Comment by vincent178** [comment]
@jsumners Yes, I saw it, thanks.

I'm heading to bed now, if there's no objection, I could send a PR to `fastify-reply-from` tomorrow to reset the `content-length`.

**Comment by mcollina** [comment]
Node.js automatically dump the uncosumed request body when endind a
response.
Il giorno dom 27 mag 2018 alle 16:01 James Sumners <notifications@github.com>
ha scritto:

**Comment by mcollina** [comment]
Go ahead and send a PR.
Il giorno dom 27 mag 2018 alle 16:10 Matteo Collina <
matteo.collina@gmail.com> ha scritto:

**Comment by vincent178** [comment]
Here's the PR: https://github.com/fastify/fastify-reply-from/pull/12/files
@jsumners @mcollina

**Comment by mcollina** [comment]
Thanks for the PR! Released as 0.4.3.
